src/Microsoft.AspNetCore.JsonPatch/Adapters/AdapterFactory.cs

```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.JsonPatch.Internal;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.AspNetCore.JsonPatch.Adapters
+{
+    /// <summary>
+    /// The default AdapterFactory to be used for resolving <see cref="IAdapter"/>.
+    /// </summary>
+    public class AdapterFactory : IAdapterFactory
+    {
+        /// <inheritdoc />
+        public virtual IAdapter Create(object target, IContractResolver contractResolver)
```

Comment:
How about just passing in the `JsonContract` since that's all the `contractResolver` will be used for in this context?

Reply:
Either way.  The `JsonContract` must match the target type passed in.  Since the code doesn't already have a `JsonContract` we can trust the developer to properly create it and pass it in, or just create it in the function.  The original `ObjectVisitor.SelectAdapter` method generated the `JsonContract` in the method, so I followed the same approach.  We can create it in the `ObjectVisitor.SelectAdapter` but if somebody tries to use an `AdapterFactory` outside the current scope they have to do it correctly.  This seemed less error prone to me, but if you would prefer it the other way, it is an easy change.  Let me know.

Reply:
I prefer passing in `JsonContract` instead of `contractResolver`.  I'm assuming `AdapterFactory` won't be used in an unforeseen way.  This way the `JsonContract` will be gotten once instead of by every `IAdapterFactory` implementation, which seems less error-prone to me.

Reply:
I don't think it can reuse the `JsonContract` because that is specific to the target for that given `SelectAdapter` call.  For example, if you look at `ObjectVisitor.TryVisit' you will see that we are passing in a different target as we traverse the path.  That results in a different 'JsonContract' each time.  So we aren't getting any performance gains of reuse.  Ultimately it comes down to putting that line of code in 'ObjectVisitor.SelectAdapter' or in 'AdapterFactory.Create'.  If it goes into `SelectAdapter` we are potentially limiting a future extension of the 'AdapterFactory' by only providing the 'JsonContract'.  Am I missing something?

Reply:
My words "gotten once" were poorly chosen.  I meant to suggest that the line of code for getting the `JsonContract` be in only one place in the library (`ObjectVisitor.SelectAdapter`), instead of in every `IAdapterFactory` implementation.  My idea is to keep `AdapterFactory.Create` tightly scoped to it's intended purpose.  Maybe you have a potential scenario in mind that I haven't considered?

Reply:
Just wondering how people might try and use the AdapterFactory for other purposes and having that code embedded gives them more control and makes it less likely for somebody to pass in the wrong JsonContract.  I think it comes down to 6 or a half dozen, so I can go ahead and make the change.

Reply:
This discussion got me thinking: here I am arguing about `contractResolver` vs. `jsonContract` but none of my extension scenarios require either.  So now I have a different suggestion.  Let's pass `IAdapterFactory.Create` neither `contractResolver` nor `jsonContract`.  As you pointed out, we shouldn't preclude anyone from using them.  But we're not!  For `AdapterFactory` for example, its constructor can accept `contractResolver`.

Reply:
I see the difference of where we're coming from.  You're arguing against redundancy of the data passed to `Create`.  I'm arguing against redundancy of the code that generates that data.  I think I found a compromise that addresses both concerns, and leads to a better API too.  I posted my new suggestion on your updated code.